### PR TITLE
osu-lazer: init at 2020.714.0

### PR DESCRIPTION
**Pull request by oxalica**
###### Motivation for this change
Add a new package `osu-lazer`, a rhythm game.
Home page: https://osu.ppy.sh
Repository (osu!lazer): https://github.com/ppy/osu

~It seems to have darwin releases, but I'm not familiar with it. Anyone help?~
Now it is built from source, targeting only x86_64-linux.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
  - 441.2M
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/games/osu/default.nix [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+  profile = ''
+    export LD_LIBRARY_PATH="$LD_LIBRARY_PATH''${LD_LIBRARY_PATH:+:}${icu}/lib"
+  '';
+
+  meta = with lib; {
+    description = "Rhythm is just a click away";
+    homepage = "https://osu.ppy.sh";
+    license = licenses.mit;
+    maintainers = with maintainers; [ oxalicc ];
```
Comment (cole-h) [inline pkgs/games/osu/default.nix:26]:
```suggestion
    maintainers = with maintainers; [ oxalica ];
```

One letter off, and borg would have been happy :P

Comment (oxalica) [reply pkgs/games/osu/default.nix]: Oops, it's not evaluated when building. :(

**Review by cole-h — COMMENTED** [review]
It runs just fine, but I get no sound, and the following is printed:

```
ALSA lib dlmisc.c:285:(snd1_dlobj_cache_get) Cannot open shared library /nix/store/gac5607idcq82hmlr2filyjlygizwgky-alsa-lib-1.1.9/lib/alsa-lib/libasound_module_pcm_pulse.so ((null): /nix/store/gac5607idcq82hmlr2filyjlygizwgky-alsa-lib-1.1.9/lib/alsa-lib/libasound_module_pcm_pulse.so: cannot open shared object file: No such file or directory)
```

Could just be a quirk of me not being on NixOS, though. I'd recommend finding someone else to verify or refute my claim.

**Comment by symphorien** [comment]
It's easier said than done, but did you try building from source ? Apparently there are utilities to build dot net projects in nixpkgs, see an example with keepass: https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/misc/keepass/default.nix

**Comment by dramforever** [comment]
It seems that osu!lazer's code is licensed under MIT [as shown here][osu-license], but the assets are licensed under CC-BY-NC 4.0 [as shown here][osu-resources-license]. I don't know where that puts the binary releases, and I also don't know how to describe 'part MIT, part CC', but it's something to take note of.

[osu-license]: https://github.com/ppy/osu#licence
[osu-resources-license]: https://github.com/ppy/osu-resources#licence

**Comment by oxalica** [comment]
@symphorien 
Simply using `buildDotnetPackage` results in build failure as below.

> The default XML namespace of the project must be the MSBuild XML namespace. If the project is authored in the MSBuild 2003 format, please add xmlns="http://schemas.microsoft.com/developer/msbuild/2003" to the <Project> element. If the project has been authored in the old 1.0 or 1.2 format, please convert it to MSBuild 2003 format.

Seems the format of project file is too new for `mono` `xbuild`. And if you add `xmlns` as the description, it will produce more errors about unrecognized attributes.

**Comment by symphorien** [comment]
Ah ok. Thanks for trying anyway.

**Comment by oxalica** [comment]
- Re-package to build from source with `dotnet`.
- Fix license.
- Bump to 2020.413.0
- Rename package name to `osu-lazer` (following [AUR](https://aur.archlinux.org/packages/osu-lazer/))

It works on nixos but still crash in my lubuntu VM. It seems to be something wrong with X11 and/or OpenGL.
Here's the error log:

```text
Unhandled exception. osuTK.Graphics.GraphicsContextException: Invalid window
   at osuTK.Platform.SDL2.Sdl2GraphicsContext..ctor(GraphicsMode mode, IWindowInfo win, IGraphicsContext shareContext, Int32 major, Int32 minor, GraphicsContextFlags flags)
   at osuTK.Platform.SDL2.Sdl2Factory.CreateGLContext(GraphicsMode mode, IWindowInfo window, IGraphicsContext shareContext, Boolean directRendering, Int32 major, Int32 minor, GraphicsContextFlags flags)
   at osuTK.Graphics.GraphicsContext..ctor(GraphicsMode mode, IWindowInfo window, IGraphicsContext shareContext, Int32 major, Int32 minor, GraphicsContextFlags flags)
   at osuTK.Graphics.GraphicsContext..ctor(GraphicsMode mode, IWindowInfo window, Int32 major, Int32 minor, GraphicsContextFlags flags)
   at osuTK.GameWindow..ctor(Int32 width, Int32 height, GraphicsMode mode, String title, GameWindowFlags options, DisplayDevice device, Int32 major, Int32 minor, GraphicsContextFlags flags, IGraphicsContext sharedContext, Boolean isSingleThreaded)
   at osuTK.GameWindow..ctor(Int32 width, Int32 height, GraphicsMode mode, String title, GameWindowFlags options, DisplayDevice device, Int32 major, Int32 minor, GraphicsContextFlags flags, IGraphicsContext sharedContext)
   at osuTK.GameWindow..ctor(Int32 width, Int32 height, GraphicsMode mode, String title, GameWindowFlags options, DisplayDevice device, Int32 major, Int32 minor, GraphicsContextFlags flags)
   at osuTK.GameWindow..ctor(Int32 width, Int32 height, GraphicsMode mode, String title, GameWindowFlags options, DisplayDevice device)
   at osuTK.GameWindow..ctor(Int32 width, Int32 height, GraphicsMode mode)
   at osu.Framework.Platform.GameWindow..ctor(Int32 width, Int32 height)
   at osu.Framework.Platform.DesktopGameWindow..ctor()
   at osu.Framework.Platform.Linux.LinuxGameWindow..ctor()
   at osu.Framework.Platform.Linux.LinuxGameHost.CreateWindow()
   at osu.Framework.Platform.GameHost.Run(Game game)
   at osu.Desktop.Program.Main(String[] args) in /build/source/osu.Desktop/Program.cs:line 57
```

**Comment by symphorien** [comment]
OpenGL is impure and sometimes does not work on non NixOS: see https://github.com/guibou/nixGL for a possible solution.

**Comment by oxalica** [comment]
Bumped to 2020.421.0 with some tiny fixes.

> OpenGL is impure and sometimes does not work on non NixOS: see https://github.com/guibou/nixGL for a possible solution.

I checked nixGL and it run successfully in my lubuntu VM, without display or audio problem.
Well, I don't think the OpenGL issue should be solved by package itself, since hard-coding drivers is a bad idea.

@cole-h What do you think?

File: pkgs/games/osu-lazer/default.nix [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+      url = "https://www.nuget.org/api/v2/package/${name}/${version}";
+      inherit sha256;
+    };
+  });
+
+  configurePhase = ''
+    runHook preConfigure
+
+    mkdir home
+    export HOME=$PWD/home
```
Comment (cole-h) [inline pkgs/games/osu-lazer/default.nix:39]:
```suggestion
    export HOME=$(mktemp -d)
```

**Review by cole-h — CHANGES_REQUESTED** [review]
I get a build failure using `nixpkgs-review`:

```
builder for '/nix/store/1nrhdv1cgyh9r8dqmdsss9z82f3na340-osu-lazer-2020.421.0.drv' failed with exit code 2; last 10 log lines:
      0 Warning(s)
      0 Error(s)

  Time Elapsed 00:00:14.51
  installing
  Microsoft (R) Build Engine version 16.4.0+e901037fe for .NET Core
  Copyright (C) Microsoft Corporation. All rights reserved.

    osu.Desktop -> /nix/store/ghp1mgw009wnzdia1swg5piwmhba5llx-osu-lazer-2020.421.0/lib/osu/
  /nix/store/shr8mc5hx0vasv92b307mzymj2b5fxxl-stdenv-linux/setup: eval: line 1307: syntax error near unexpected token `('
cannot build derivation '/nix/store/ysx6famjfkaqn93dnb9c99i0lpbzbz07-env.drv': 1 dependencies couldn't be built
[288 built (1 failed), 10 copied (60.1 MiB), 38.4 MiB DL]
error: build of '/nix/store/ysx6famjfkaqn93dnb9c99i0lpbzbz07-env.drv' failed
https://github.com/NixOS/nixpkgs/pull/85021
1 package failed to build:
osu-lazer
```

**Comment by oxalica** [comment]
Ping @cole-h

**Review by cole-h — COMMENTED** [review]
I can't really approve this because I still don't get audio on my Arch+Nix system. But it builds fine and runs fine apart from that.

```
ALSA lib dlmisc.c:283:(snd_dlobj_cache_get0) Cannot open shared library /nix/store/i58krnmjdd5aq47jxxljz99csk91v1s8-alsa-lib-1.2.2/lib/alsa-lib/libasound_module_pcm_pulse.so ((null): /nix/store/i58krnmjdd5aq47jxxljz99csk91v1s8-alsa-lib-1.2.2/lib/alsa-lib/libasound_module_pcm_pulse.so: cannot open shared object file: No such file or directory)
```

---

LGTM, otherwise.

```
[5 built, 0.0 MiB DL]
https://github.com/NixOS/nixpkgs/pull/85021
1 package built:
osu-lazer
```

---

EDIT: Also, please update the PR title again ;)

**Comment by oxalica** [comment]
Rebased and bumped to 2020.508.2

@cole-h I've setup-ed an clean Arch Linux virtual machine but cannot reproduce the ALSA error. **Audio and video both works fine.** Is `LD_LIBRARY_PATH` set in your environment? (Mine is empty)

Screenshot_20200510_165931

Installed packages in my test environment:
```
> pacman -Qet
base 2-2
dhcpcd 9.0.2-1
gpicview 0.2.5-4
grub 2:2.04-6
lxappearance-obconf 0.2.3-2
lxde-common 0.99.2-2
lxdm 0.5.3-6
lxhotkey 0.1.0-2
lxinput 0.3.5-2
lxlauncher 0.2.5-3
lxmusic 0.4.7-2
lxpanel 0.10.0-1
lxrandr 0.3.2-1
lxsession 1:0.5.5-1
lxtask 0.1.9-1
lxterminal 0.3.2-2
openssh 8.2p1-3
pcmanfm 1.3.1-1
rsync 3.1.3-3
screenfetch 3.9.1-1
sudo 1.8.31.p1-1
vim 8.2.0717-2
```

**Comment by oxalica** [comment]
Ping @cole-h

**Review by cole-h — APPROVED** [review]
Diff LGTM, binary runs. Still no sound, which means it's a problem with my crufty machine (and thus not a blocker for merging this).

```
[4 built, 7 copied (700.8 MiB), 171.9 MiB DL]
https://github.com/NixOS/nixpkgs/pull/85021
1 package built:
osu-lazer
```

**Comment by oxalica** [comment]
@cole-h What is blocking this now?

**Comment by cole-h** [comment]
> @cole-h What is blocking this now?

Waiting on an interested party that has merge capabilities and is willing to exercise that ability.

**Comment by oxalica** [comment]
- Squash and bumped to 2020.707.0
- Add `passthru.updateScript`

**Review by D4ndellion — APPROVED** [review]
Built and tested

2020.714.0 is out now too so needs a bump

**Comment by skykanin** [comment]
Hey, I made a PR for this back in May, but was never able to get it working properly so I closed it in favour of this one. Built this on my NixOS machine at it worked perfectly including no audio issues. Other than needing a bump to version 2020.714.0 this looks good to me. :heart:

**Review by skykanin — APPROVED** [review]

**Comment by skykanin** [comment]
Also the PR title needs to be updated to reflect version bump

**Comment by catern** [comment]
I'm mentioning this here because maybe it's a packaging bug, it looks like none of the networked functionality of osu is working - downloading beatmaps, making an osu account, etc. Anyone else observing this? I'm guessing this might be because of the disabled tamper-detection?

Kind of makes it unusable since you need to have osu to even make an osu account :)

**Pull request merged**